### PR TITLE
Optimise generated code

### PR DESCRIPTION
**Pull request by miklos1**
A few optimisations that help to achieve assembly performance on par with our old FFC on affine cells:
 * Accumulating directly into the output tensor instead of using a temporary and then later copying into the output tensor.
 * Alternative mode of handling mixed element coefficients in interior facet integrals: assemble ListTensor instead "manually" generating COFFEE loops.
 * This alternative way is enabled on the coordinates on simplex cells. Combined with the previous optimisation expanding the evaluation of cellwise constant coefficients, this gives "code snippet"-equivalent code even in interior facet integrals.
 * Don't expand quadrature loops.

### CODE REVIEW

File: tsfc/driver.py [hunk 1]
```diff
@@ -100,7 +100,10 @@ def compile_integral(idata, fd, prefix, parameters):
 
     mesh = idata.domain
     coordinates = fem.coordinate_coefficient(mesh)
-    funarg, prepare_, expression = prepare_coefficient(integral_type, coordinates, "coords")
+    if mesh.ufl_cell().cellname() in ["interval", "triangle", "tetrahedron"]:
+        funarg, prepare_, expression = prepare_coefficient(integral_type, coordinates, "coords", mode='list_tensor')
```
Comment (wence-) [inline tsfc/driver.py]: If I understand the description correctly, you want to do this for affine coordinates?  But this is not the correct check for that.

File: tsfc/driver.py [hunk 2]
```diff
@@ -241,7 +243,11 @@ def compile_integral(idata, fd, prefix, parameters):
     return kernel
 
 
-def prepare_coefficient(integral_type, coefficient, name):
+def prepare_coefficient(integral_type, coefficient, name, mode=None):
+    if mode is None:
+        mode = 'manual_loop'
+
+    assert mode in ['manual_loop', 'list_tensor']
```
Comment (wence-) [inline tsfc/driver.py]: Is there a reason not to always use one over the other (and when should one make the choice?)

File: tsfc/driver.py [hunk 3]
```diff
@@ -291,30 +297,54 @@ def prepare_coefficient(integral_type, coefficient, name):
         return funarg, [], expression
 
     # Interior facet integral + mixed / vector element
-    name_ = name + "_"
-    shape = (2, fiat_element.space_dimension())
```
Comment (wence-) [inline tsfc/driver.py]: Maybe a short comment on what this packing is doing?  And why there are two options?

**Comment by miklos1** [comment]
@wence-: I've added clarifications and a fix, please have a look.

**Comment by wence-** [comment]
Yes, thanks.

**Pull request merged**